rpc/handler.go

```diff
@@ -91,6 +91,77 @@ func newHandler(connCtx context.Context, conn jsonWriter, idgen func() ID, reg *
 	return h
 }
 
+// batchCallBuffer manages in progress calls and their responses during a batch call.
+// The calls need to be synchronized between the processing and timeout-triggering goroutines.
+type batchCallBuffer struct {
+	mutex   sync.Mutex
+	calls   []*jsonrpcMessage
+	answers []*jsonrpcMessage
+	wrote   bool
+}
+
+func (b *batchCallBuffer) timeout(ctx context.Context, conn jsonWriter) {
+	b.mutex.Lock()
+	defer b.mutex.Unlock()
+
+	for _, msg := range b.calls {
+		if !msg.isNotification() {
+			b.answers = append(b.answers, msg.errorResponse(&internalServerError{errcodeTimeout, errMsgTimeout}))
+		}
+	}
+
+	b.write(ctx, conn)
+}
+
+func (b *batchCallBuffer) popCall() *jsonrpcMessage {
```

Comment:
Since this method doesn't _actually_ `pop`, perhaps `peekCall` would a more apt name?